File: tests/Hedgehog.Autogen.Tests/GenericGenTests.fs
module Hedgehog.AutoGen.Tests.GenericGenTests

open System
open Xunit
open Swensen.Unquote
open Hedgehog
open Hedgehog.FSharp

type Uuid = Uuid of Guid
type Name = Name of string
type Id<'a> = Id of Guid
type Either<'a, 'b> = Left of 'a | Right of 'b
type Maybe<'a> = Just of 'a | Nothing

type OuterRecord = { Value: Maybe<Guid> }

type OuterClass(value: Maybe<Guid>) =
    member val Value = value with get, set


type GenericTestGenerators =

  // Test that we can override the "default" generator for a type.
  static member Guid() =
    Gen.constant Guid.Empty

  // A generator for Id<'a> to test phantom generic type.
  static member Id<'a>(gen : Gen<Guid>) : Gen<Id<'a>> = gen |> Gen.map Id

  // A generator for some simple value to test a generator without parameters.
  static member UuidGen() : Gen<Uuid> = GenericTestGenerators.Guid() |> Gen.map Uuid

  // A generator for some simple value to test a generator with parameters.
  static member NameGen(gen: Gen<string>) : Gen<Name> =
    gen |> Gen.map (fun x -> Name ("Name: " + x))

  // A generator for Maybe<'a> to test union type with one generic type constructor.
  static member AlwaysJust<'a>(genA: Gen<'a>) : Gen<Maybe<'a>> = genA |> Gen.map Just

  // A generator for Either<'a, 'b> to test union type with multiple type constructors.
  static member AlwaysLeft<'a, 'b>(genB: Gen<'b>, genA: Gen<'a>) : Gen<Either<'a, 'b>> =
        genA |> Gen.map Left

  static member MaybeStringSpecific(): Gen<Maybe<string>> =
        Gen.constant (Just "Specific String")

  static member EitherIntStringSpecific<'a>(): Gen<Either<'a, string>> =
        Gen.constant (Right "Specific String")

let checkWith tests = PropertyConfig.defaults |> PropertyConfig.withTests tests |> Property.checkWith

let isCustomGuid (guid: Guid) = guid.ToByteArray()[..3] |> Array.forall ((=) 0uy)

[<Fact>]
let ``should generate value with phantom generic type - Id<'a>``() =
   let config = AutoGenConfig.defaults |> AutoGenConfig.addGenerators<GenericTestGenerators>
   checkWith 100<tests> <| property {
        let! x = Gen.autoWith<Id<string>> config
        test <@ x |> function Id a -> isCustomGuid a @>
    }

[<Fact>]
let ``should generate generic value for union type - Either<'a, 'b>``() =
   let config = AutoGenConfig.defaults |> AutoGenConfig.addGenerators<GenericTestGenerators>
   checkWith 100<tests> <| property {
        let! x = Gen.autoWith<Either<int, bool>> config
        test <@ x |> function Left _ -> true | _ -> false @>
    }

[<Fact>]
let ``should generate generic value for union type - Maybe<'a>``() =
   let config = AutoGenConfig.defaults |> AutoGenConfig.addGenerators<GenericTestGenerators>
   checkWith 100<tests> <| property {
        let! x = Gen.autoWith<Maybe<string>> config
        test <@ x |> function Just _ -> true  | _ -> false @>
    }

[<Fact>]
let ``should generate value using a generator without parameters: Uuid``() =
   let config = AutoGenConfig.defaults |> AutoGenConfig.addGenerators<GenericTestGenerators>
   checkWith 100<tests> <| property {
        let! x = Gen.autoWith<Maybe<Uuid>> config
        test <@ x |> function Just (Uuid x) -> isCustomGuid x  | _ -> failwith "todo"@>
    }

[<Fact>]
let ``should generate value using a generator with parameters: Name``() =
   let config = AutoGenConfig.defaults |> AutoGenConfig.addGenerators<GenericTestGenerators>
   checkWith 100<tests> <| property {
        let! x = Gen.autoWith<Name> config
        test <@ x |> function Name x -> x.StartsWith("Name: ") @>
    }

[<Fact>]
let ``should generate outer FSharp record with generic type inside``() =
   let config = AutoGenConfig.defaults |> AutoGenConfig.addGenerators<GenericTestGenerators>
   checkWith 100<tests> <| property {
        let! x = Gen.autoWith<OuterRecord> config
        test <@ x |> function { Value = Just x } -> isCustomGuid x | _ -> false @>
    }

[<Fact>]
let ``should generate outer class with generic type inside``() =
   let config = AutoGenConfig.defaults |> AutoGenConfig.addGenerators<GenericTestGenerators>
   checkWith 100<tests> <| property {
        let! x = Gen.autoWith<OuterClass> config
        test <@ x |> function cls -> match cls.Value with Just v -> isCustomGuid v | _ -> false @>
    }

[<Fact>]
let ``should use most specific generators``() =
   let config = AutoGenConfig.defaults |> AutoGenConfig.addGenerators<GenericTestGenerators>
   checkWith 100<tests> <| property {
        let! x = Gen.autoWith<Maybe<string>> config
        test <@ x = Just "Specific String" @>

        let! y = Gen.autoWith<Either<int, string>> config
        test <@ y = Right "Specific String" @>
    }

File: tests/Hedgehog.Autogen.Tests.CSharp/GenericGenTests.cs
using System.Collections.Immutable;
using Xunit;
using Hedgehog.Linq;
using Range = Hedgehog.Linq.Range;

namespace Hedgehog.AutoGen.Linq.Tests;

public sealed record Uuid(Guid Value);

public sealed record Name(string Value);

public sealed record Id<T>(Guid Value);

public abstract record Either<TLeft, TRight>
{
  public sealed record Left(TLeft Value) : Either<TLeft, TRight>;

  public sealed record Right(TRight Value) : Either<TLeft, TRight>;
}

public abstract record Maybe<T>
{
  public sealed record Just(T Value) : Maybe<T>;

  public sealed record Nothing : Maybe<T>;
}

public sealed record OuterRecord(Maybe<Guid> Value);

public sealed class OuterClass
{
  public OuterClass(Maybe<Guid> value) => Value = value;
  public Maybe<Guid> Value { get; set; }
}

public sealed record RecursiveRec(Maybe<RecursiveRec> Value);

public sealed class GenericTestGenerators
{
  public static Gen<Guid> Guid() =>
    Gen.Byte(Range.ConstantBoundedByte())
      .Array(Range.Singleton(12))
      .Select(bytes => new byte[4].Concat(bytes).ToArray())
      .Select(bytes => new Guid(bytes));

  public static Gen<Id<T>> IdGen<T>(Gen<Guid> gen) =>
    gen.Select(value => new Id<T>(value));

  public static Gen<Uuid> UuidGen() =>
    Guid().Select(value => new Uuid(value));

  public static Gen<Name> NameGen(Gen<string> gen) =>
    gen.Select(value => new Name("Name: " + value));

  public static Gen<Maybe<T>> AlwaysJust<T>(AutoGenContext context, Gen<T> gen) =>
    context.CanRecurse
      ? gen.Select(Maybe<T> (value) => new Maybe<T>.Just(value))
      : Gen.Constant<Maybe<T>>(new Maybe<T>.Nothing());

  public static Gen<Either<TLeft, TRight>> AlwaysLeft<TLeft, TRight>(Gen<TRight> genB, Gen<TLeft> genA) =>
    genA.Select(Either<TLeft, TRight> (value) => new Either<TLeft, TRight>.Left(value));
}

public class GenericGenTests
{
  private static bool IsCustomGuid(Guid guid) =>
    new Span<byte>(guid.ToByteArray(), 0, 4).ToArray().All(b => b == 0);

  [Fact]
  public void ShouldGenerateRecursiveRecords()
  {
    var config = AutoGenConfig.Defaults.AddGenerators<GenericTestGenerators>();
    var prop = from x in Gen.AutoWith<RecursiveRec>(config).ForAll()
               select x != null;

    prop.Check();
  }

  [Fact]
  public void ShouldGenerateValueWithPhantomGenericType_Id()
  {
    var config = AutoGenConfig.Defaults.AddGenerators<GenericTestGenerators>();
    var prop = from x in Gen.AutoWith<Id<string>>(config).ForAll()
               select IsCustomGuid(x.Value);

    prop.Check();
  }

  [Fact]
  public void ShouldGenerateGenericValueForUnionType_Either()
  {
    var config = AutoGenConfig.Defaults.AddGenerators<GenericTestGenerators>();
    var prop = from x in Gen.AutoWith<Either<int, string>>(config).ForAll()
               select x is Either<int, string>.Left;
    prop.Check();
  }

  [Fact]
  public void ShouldGenerateGenericValueForUnionType_Maybe()
  {
    var config = AutoGenConfig.Defaults.AddGenerators<GenericTestGenerators>();
    var prop = from x in Gen.AutoWith<Maybe<string>>(config).ForAll()
               select x is Maybe<string>.Just;
    prop.Check();
  }

  [Fact]
  public void ShouldGenerateValueUsingGeneratorWithoutParameters_Uuid()
  {
    var config = AutoGenConfig.Defaults.AddGenerators<GenericTestGenerators>();
    var prop = from x in Gen.AutoWith<Uuid>(config).ForAll()
               select IsCustomGuid(x.Value);
    prop.Check();
  }

  [Fact]
  public void ShouldGenerateValueUsingGeneratorWithParameters_Name()
  {
    var config = AutoGenConfig.Defaults.AddGenerators<GenericTestGenerators>();
    var prop = from x in Gen.AutoWith<Name>(config).ForAll()
               select x.Value.StartsWith("Name: ");
    prop.Check();
  }

  [Fact]
  public void ShouldGenerateOuterFSharpRecordWithGenericTypeInside()
  {
    var config = AutoGenConfig.Defaults.AddGenerators<GenericTestGenerators>();
    var prop = from x in Gen.AutoWith<OuterRecord>(config).ForAll()
      select x.Value switch
      {
        Maybe<Guid>.Just(var v) => IsCustomGuid(v),
        Maybe<Guid>.Nothing => false,
        _ => throw new InvalidOperationException("C# cannot do exhaustive matching")
      };

    prop.Check();
  }

  [Fact]
  public void ShouldGenerateOuterClassWithGenericTypeInside()
  {
    var config = AutoGenConfig.Defaults.AddGenerators<GenericTestGenerators>();
    var prop = from x in Gen.AutoWith<OuterClass>(config).ForAll()
      select x.Value switch
      {
        Maybe<Guid>.Just(var v) => IsCustomGuid(v),
        Maybe<Guid>.Nothing => false,
        _ => throw new InvalidOperationException("C# cannot do exhaustive matching")
      };
    prop.Check();
  }

  [Fact]
  public void ShouldGenerateImmutableListUsingAutoGenConfigParameter()
  {
    var config = AutoGenConfig.Defaults
      .SetCollectionRange(Range.Singleton(7))
      .AddGenerators<GenericTestGenerators>();

    var prop = from x in Gen.AutoWith<ImmutableList<int>>(config).ForAll()
               select x.Count == 7;

    prop.Check();

    var propString = from x in Gen.AutoWith<ImmutableList<string>>(config).ForAll()
                     select x.Count == 7;

    propString.Check();
  }
}
